### ISSUE TITLE
Error creating order in Bitso

### ISSUE THREAD

**Issue by rubobaquero** [ccxt/ccxt#525]
I receive an error when i try to create an order with Bitso:

Example code:

```
$exchange = new \ccxt\bitso(['apiKey'=>'XXXX','secret'=>'XXXX']);
$exchange->verbose = true;
$exchange->timeout = 10000;
$exchange->load_markets();
$balance = $exchange->fetch_balance(); // This works OK, so it isn't an auth problem

$volume = 20;
$price = 0.00002950;
$order = $exchange->create_order("XRP/BTC", "limit", "buy", $volume, $price);
```
Error:
```
PHP Warning:  array_key_exists() expects parameter 2 to be array, null given in /XXXX/ccxt/php/bitso.php on line 227

Warning: array_key_exists() expects parameter 2 to be array, null given in /XXXX/ccxt/php/bitso.php on line 227
PHP Fatal error:  Uncaught ccxt\ExchangeError: bitso null in /XXXX/ccxt/php/bitso.php:230
Stack trace:
#0 [internal function]: ccxt\bitso->request('orders', 'private', 'POST', Array)
#1 /XXXX/ccxt/php/base/Exchange.php(545): call_user_func(Array, 'orders', 'private', 'POST', Array)
#2 [internal function]: ccxt\Exchange->ccxt\{closure}(Array)
```

And this is the request that is making the code when activating verbose (I have modified the Authorization header to avoid security problems):
```
Request:
Array
(
    [0] => POST
    [1] => https://api.bitso.com/v3/orders
    [2] => Array
        (
            [0] => Authorization: Bitso XXX:XXX:XXX
            [User-Agent] => ccxt/1.9.282 (+https://github.com/ccxt-dev/ccxt) PHP/7.0.14
        )

    [3] => {"book":"xrp_btc","side":"buy","type":"limit","major":20,"price":2.95e-5}
)
```

**Comment by kroitor** [comment]
Hi, @rubobaquero ! Thanks for reporting this, I will take a look into it and will get back to with a fix for this.

**Comment by kroitor** [comment]
@rubobaquero ok, I see several issues here:

A. In your code I see that you are not respecting the rate limit of bitso... So, if you remove the fetch_balance call, your code should work right away.

B. You should not do more than one request per .rateLimit by default. Or your should set `.enableRateLimit = true` on the exchange instance or in the constructor if you want it to sleep for that delay time for you. If you enable the built-in rate limiter, you can do fetch_balance and then, sequentially, the create_order, it will not exceed the rate limits.

C. Alternatively, you can also override the nonce function if you want to do requests more frequently and at the same time avoid nonce conflicts. Just use the milliseconds-nonce. More about it here:

https://github.com/kroitor/ccxt/wiki/Manual#troubleshooting
https://github.com/kroitor/ccxt/wiki/Manual#overriding-the-nonce
https://github.com/kroitor/ccxt/wiki/Manual#authentication

D. Apart from that I see that price is formatted in scientific notation, though I'm not sure whether this is an issue with the verbose output, or with the JSON value itself. Should actually work normally, as the same code is used in other exchanges as well.

I suggest trying it in the above order, step by step. First, try A (remove fetch_balance) and report if it works. The try B (enable rate limiter, restore the fetch_balance call, and report). Then C... If you have a difficulty doing any of these steps, let us know.

**Comment by rubobaquero** [comment]
Hi kroitor,
I tried with your suggestion of removing fetch_balance and received the same error:

Example
```
$exchange_name = "bitso";
$exchange = new \ccxt\bitso(Config::$data['exchanges'][$exchange_name]);
$exchange->verbose = true;
$exchange->timeout = 10000;
$exchange->load_markets();

$volume = 20;
$price = 0.00002950;
$order = $exchange->create_order("XRP/BTC", "limit", "buy", $volume, $price);
```

Error
```
Warning: array_key_exists() expects parameter 2 to be array, null given in /XXXX/ccxt/php/bitso.php on line 227
PHP Fatal error:  Uncaught ccxt\ExchangeError: bitso null in /XXXX/ccxt/php/bitso.php:230
Stack trace:
#0 [internal function]: ccxt\bitso->request('orders', 'private', 'POST', Array)
#1 /XXXX/ccxt/php/base/Exchange.php(545): call_user_func(Array, 'orders', 'private', 'POST', Array)
#2 [internal function]: ccxt\Exchange->ccxt\{closure}(Array)
#3 /XXXX/ccxt/php/base/Exchange.php(1298): call_user_func_array(Object(Closure), Array)
#4 /XXXX/ccxt/php/bitso.php(195): ccxt\Exchange->__call('privatePostOrde...', Array)
#5 /XXXX/trading/test_bitso.php(28): ccxt\bitso->create_order('XRP/BTC', 'limit', 'buy', 20, 2.95E-05)
#6 {main}
  thrown in /XXXX/ccxt/php/bitso.php on line 230

Fatal error: Uncaught ccxt\ExchangeError: bitso null in /XXXX/ccxt/php/bitso.php:230
Stack trace:
#0 [internal function]: ccxt\bitso->request('orders', 'private', 'POST', Array)
#1 /XXXX/ccxt/php/base/Exchange.php(545): call_user_func(Array, 'orders', 'private', 'POST', Array)
#2 [internal function]: ccxt\Exchange->ccxt\{closure}(Array)
#3 /XXXX/ccxt/php/base/Exchange.php(1298): call_user_func_array(Object(Closure), Array)
#4 /XXXX/ccxt/php/bitso.php(195): ccxt\Exchange->__call('privatePostOrde...', Array)
#5 /XXXX/trading/test_bitso.php(28): ccxt\bitso->create_order('XRP/BTC', 'limit', 'buy', 20, 2.95E-05)
#6 {main}
  thrown in /XXXX/ccxt/php/bitso.php on line 230
```

Request:
```
Array
(
    [0] => POST
    [1] => https://api.bitso.com/v3/orders
    [2] => Array
        (
            [0] => Authorization: Bitso XXX:XXX:XXX
            [User-Agent] => ccxt/1.9.282 (+https://github.com/ccxt-dev/ccxt) PHP/7.0.14
        )

    [3] => {"book":"xrp_btc","side":"buy","type":"limit","major":20,"price":2.95e-5}
)
```
If I try the same with Bittrex or Poloniex for example it works. My Version is 1.10.49.

**Comment by rubobaquero** [comment]
Same problem overriding the default nonce:
```
class mybitso extends \ccxt\bitso {
	public function __construct ($options = array ()) {
		parent::__construct (array_merge (array ('i' => 1), $options));
	}
	public function nonce () {
		return $this->milliseconds ();
	}
}

$exchange = new mybitso(['apiKey'=>'XXXX','secret'=>'XXXX']);
$exchange->verbose = true;
$exchange->timeout = 10000;
$exchange->load_markets();

$volume = 20;
$price = 0.00002950;
$order = $exchange->create_order("XRP/BTC", "limit", "buy", $volume, $price);
```

**Comment by rubobaquero** [comment]
And same problem enabling rate limit:

```
$exchange = new \ccxt\bitso(['apiKey'=>'XXXX','secret'=>'XXXX']);
$exchange->verbose = true;
$exchange->timeout = 10000;
$exchange->enableRateLimit = true;
$exchange->load_markets();

$balances = $exchange->fetch_balance();

$volume = 20;
$price = 0.00002950;
$order = $exchange->create_order("XRP/BTC", "limit", "buy", $volume, $price);
```

Just for curiosity, i have dumped the response from bitso obtained in $exchange->fetch() method when trying to create the order:

`Disallowed Key Characters`

Probably the problem is the scientific notation?  What should i put instead? number_format($price,8) ?

**Comment by kroitor** [comment]
@rubobaquero ok, thank you for a comprehensive report! I also think this is a "D" problem now, I'll look into it and will get back to you with a fix for it asap. Thx again!

**Comment by kroitor** [comment]
@rubobaquero this issue was fixed in 1.10.54. Let us know if you have any remaining difficulties with it. Thx!
